File: RobustDemo/app/src/main/java/com/meta/robustdemo/MainActivity.java
package com.meta.robustdemo;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Toast;

import com.meituan.robust.PatchExecutor;
import com.meta.robustdemo.robust.PatchManipulateImp;
import com.meta.robustdemo.robust.PermissionUtils;
import com.meta.robustdemo.robust.RobustCallBackSample;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        findViewById(R.id.btn_patch).setOnClickListener(this);

        findViewById(R.id.btn_to_second_activity).setOnClickListener(this);

        findViewById(R.id.btn_bug).setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.btn_patch: {
                if (isGrantSDCardReadPermission()) {
                    runRobust();
                } else {
                    requestPermission();
                }
            }
            break;
            case R.id.btn_to_second_activity: {
                startActivity(new Intent(v.getContext(), SecondActivity.class));
            }
            break;
            case R.id.btn_bug: {
                Toast.makeText(MainActivity.this, "main bug fixed222.", Toast.LENGTH_SHORT)
                        .show();
            }
            break;
            default:
                break;
        }
    }

    private boolean isGrantSDCardReadPermission() {
        return PermissionUtils.isGrantSDCardReadPermission(this);
    }

    private void requestPermission() {
        PermissionUtils.requestSDCardReadPermission(this, REQUEST_CODE_SDCARD_READ);
    }

    private static final int REQUEST_CODE_SDCARD_READ = 1;

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if (requestCode == REQUEST_CODE_SDCARD_READ) {
            handlePermissionResult();
        }
    }

    private void handlePermissionResult() {
        if (isGrantSDCardReadPermission()) {
            runRobust();
        } else {
            Toast.makeText(this, "failure because without sd card read permission", Toast.LENGTH_SHORT).show();
        }
    }

    private void runRobust() {
        new PatchExecutor(getApplicationContext(), new PatchManipulateImp(), new RobustCallBackSample()).start();
    }
}

File: ScreenShot/app/src/main/java/com/zhaopf/screenshot/MainActivity.kt
package com.zhaopf.screenshot

import android.Manifest
import android.content.pm.PackageManager
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.core.app.ActivityCompat
import androidx.core.content.PermissionChecker

class MainActivity : AppCompatActivity() {

    private lateinit var screenShotListener: ScreenShotListener
    var isHasScreenShotListener = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        screenShotListener = ScreenShotListener.getInstance(this)

        // 申请权限
        val permission = arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE)
        if (ActivityCompat.checkSelfPermission(
                        this,
                        Manifest.permission.READ_EXTERNAL_STORAGE
                ) != PackageManager.PERMISSION_GRANTED
        ) {
            ActivityCompat.requestPermissions(this, permission, 1001)
        }
    }

    override fun onRequestPermissionsResult(
            requestCode: Int,
            permissions: Array<out String>,
            grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == 1001) {
            if (grantResults[0] == PermissionChecker.PERMISSION_GRANTED) {
                customToast("权限申请成功")
            } else {
                customToast("权限申请失败")
            }
        }
    }

    override fun onResume() {
        super.onResume()
        startScreenShotListen()
    }

    override fun onPause() {
        super.onPause()
        stopScreenShotListen()
    }

    private fun startScreenShotListen() {
        if (!isHasScreenShotListener) {
            screenShotListener.setListener(object : ScreenShotListener.OnScreenShotListener {
                override fun onScreenShot(picPath: String) {
                    customToast("监听截屏成功")
                    Log.d(ScreenShotHelper.TAG, "==== picPath: $picPath")
                }
            })
            screenShotListener.startListener()
            isHasScreenShotListener = true
        }
    }

    private fun stopScreenShotListen() {
        if (isHasScreenShotListener) {
            screenShotListener.stopListener()
            isHasScreenShotListener = false
        }
    }

    private fun customToast(s: String) {
        Toast.makeText(this, s, Toast.LENGTH_SHORT).show()
    }
}
